File: day2.py
import re

def parse(line):
    m = re.match('\A(\d+)-(\d+)\s+(\w+):\s+(\w+)', line)
    return [m.group(1), m.group(2), m.group(3), m.group(4)]

def test1(min, max, letter, pw):
    cnt = pw.count(letter)
    return cnt >= int(min) and cnt <= int(max)

def test2(min, max, letter, pw):
    return (pw[int(min)-1] == letter) != (pw[int(max)-1] == letter)

lines = [parse(line) for line in open("input/input2.txt").readlines()]

# Part 1
print(len([line for line in lines if test1(*line)]))

# Part 2
print(len([line for line in lines if test2(*line)]))

File: nim/day2.nim
import re, strutils

type
    Password = tuple
        min: int
        max: int
        ch: char
        pw: string

proc parse(line: string): Password =
    var matches: array[4, string]
    if re.match(line, re"\A(\d+)-(\d+)\s+(\w+):\s+(\w+)", matches):
        return (strutils.parseInt(matches[0]), strutils.parseInt(matches[1]), 
                matches[2][0], matches[3])

proc test1(p: Password): bool =
    let cnt = strutils.count(p.pw, p.ch)
    return cnt >= p.min and cnt <= p.max

proc test2(p: Password): bool =
    return (p.pw[p.min-1] == p.ch) != (p.pw[p.max-1] == p.ch)

proc solve(filename: string, test: proc(pw: Password): bool): int =
    for line in lines filename:
        if test(parse(line)):
            result += 1


echo solve("../input/input2.txt", test1)
echo solve("../input/input2.txt", test2) 